### ISSUE TITLE
Automatically create names for syntax

### ISSUE THREAD

**Issue by MatthewDaggitt** [agda/agda#7137]
Currently it's impossible to refer to syntax declarations in a `using`, `hiding` declaration etc. because they have no name. I propose that Agda should automatically generate a name for the syntax.

As a motivating example, we've just run into a problem in v2.0 of the standard library where we converted a definition:

https://github.com/agda/agda-stdlib/blob/0817da6877aa045932221134913d8511caa78d97/src/Relation/Binary/PropositionalEquality/Core.agda#L113

into syntax:

https://github.com/agda/agda-stdlib/blob/711ac53c07e4cb3328216288c1c6206814785e81/src/Relation/Binary/Reasoning/Syntax.agda#L401

which is nearly backwards compatible unless, as PLFA [does](https://github.com/agda/agda-stdlib/issues/2298), you import the definition named.

If for every syntax declaration, Agda automatically created a name for it with misfix holes `_` substituted for the variable locations then we would be able to avoid this problem _and_ have the ability to selectively hide and use syntax as would like.

Of course syntax that uses freshly bound variables might be problematic, but I would propose fresh variables simply get zapped in the name, e.g.

```agda
syntax step-↭-⟩ x yRz x↭y = x ↭⟨ x↭y ⟩ yRz        generates    _↭⟨_⟩_
syntax step-≡-∣ x xRy     = x ≡⟨⟩ xRy             generates    _≡⟨⟩_
syntax ∄-syntax (λ x → B) = ∄[ x ] B              generates    ∄[]_
```

Possibly related issues: https://github.com/agda/agda/issues/3790, https://github.com/agda/agda/issues/3481

**Comment by andreasabel** [comment]
I am running into a related problem:  When I want to see why `_≡⟨⟩_` is in scope, I am getting the reply that it is not in scope.
(And I am running into the OP as well.)

**Comment by jamesmckinna** [comment]
I think that what I don't understand about this proposal (however much I want this as well!), is whether Agda can accept a (mixfix) name for a `syntax` declaration introduced precisely because the corresponding definition for such a name would not typecheck... (eg for having arguments out of order)

... which I understand to be the *real* reason to support them (otherwise we could simply have lots of aliasing/`renaming` directives)?

**Comment by andreasabel** [comment]
>  the corresponding definition for such a name would not typecheck... (eg for having arguments out of order)

Good point.  Indeed it would be then be only a concrete name resolved to the proper name during scope checking.  
 
> ... which I understand to be the _real_ reason to support them (otherwise we could simply have lots of aliasing/`renaming` directives)?

Yes and the fact that with syntax one can introduce one's own binder syntax.

If `syntax` declarations were not so weirdly the wrong way round (I always thought _the definiendum should be on the left_ and not on the right), we could extend them to e.g. this form:
```agda
syntax _↭⟨_⟩_ where x ↭⟨ x↭y ⟩ yRz = step-↭-⟩ x yRz x↭y 
syntax _≡⟨⟩_  where x ≡⟨⟩ xRy      = step-≡-∣ x xRy  
syntax ∄[_]_  where ∄[ x ] B       = ∄-syntax (λ x → B) 
```
Or we simply accept the OP but with `∄[_]_` in the last case which feels more consistent.

What certainly would be new is that a module can export explicitly such a syntax name and they can be explicitly imported.  (Atm, syntax for a name is attached to this name and there is no way to just import the name but not its syntax.  See also: #6918.)
And if one just imports the syntax but not its underlying name Agda would print an out-of-scope name if the syntax form cannot be recovered (are there such cases?).
